### ISSUE TITLE
Chain and iteration info for `draws_matrix`

### ISSUE THREAD

**Issue by Ozan147** [stan-dev/posterior#156]
All `draw_*` objects include the chain and iteration details except for `draws_matrix`. Would it make sense to find a way to augment `draws_matrix` with this info, perhaps by making it more similar to `draws_df`? I personally like that the remaining classes can be transform to each other seamlessly without losing any info. However `draws_matrix` collapses all draws into a single chain, making it impossible to revert or recover other classes with multiple chains.
```r
x_array <- example_draws()
# A draws_array: 100 iterations, 4 chains, and 10 variables

x_df <- as_draws_df(x_array)
# A draws_df: 100 iterations, 4 chains, and 10 variables

x_matrix <- as_draws_matrix(x_df)
# A draws_matrix: 400 draws, and 10 variables

x_df <- as_draws_df(x_matrix)
# A draws_df: 400 iterations, 1 chains, and 10 variables

x_array <- as_draws_array(x_matrix)
# A draws_array: 400 iterations, 1 chains, and 10 variables
```

**Comment by paul-buerkner** [comment]
We had thought about this when starting to develop posterior and decided to explictly let `draws_matrix` be a format without chain information, thus primarily to be used once you have already convinced yourself that your model has converged (or you have 1 chain anyway, or even just MC draws) so you could use matrices for increased convenience compared to 3D arrays. The hazzle it would need to somehow stored the chain information there (e.g., as an attribute) and then make sure this information is carried around without R dropping it is not worth it in my opinion. 

@mjskay went to some lengths to get chains infromation into the `rvar` format, but this is such a custom empty-list-with-attributes format (even though nobody sees that from the outside) that we have way more control to drop/keep stuff in certain operations anyway. However, even there we had been debating on just leaving it out. 

Accordingly, in my opinion, we should leave it as is, but I am happy to hear other opinions and be convinced otherwise.

**Comment by mjskay** [comment]
It's an interesting question. I would suspect that in most cases if you want to use an array-like format but keep chain information, it probably makes the most sense to just use draws_array instead of draws_matrix. I think the big value of draws_matrix is either (like @paul-buerkner says) when you don't care about chains anymore or when you are trying to get output that's already in that format into a draws format.

**Comment by jgabry** [comment]
I would also lean towards leaving it as is. There's definitely something nice about being able to convert between any formats and maintain all the information, but I think it's nicer in theory than in practice for the reason mentioned by others: `draws_matrix` is useful for precisely the cases where the chain information doesn't matter. 

That said, if it turns out we've underestimated the demand for keeping the chain/iteration information for `draws_matrix` then I'm definitely open to revisiting this. I think we'll find out more when we get the package on CRAN because that's when more developers will start using it in their packages (I suspect it's more likely a developer might conceivably want this than a regular user).

**Comment by jgabry** [comment]
Even though my previous comment said we shouldn't bother keeping chain info, here's an example of where it could be useful.  We just ran into an issue in CmdStanR due to the inability to convert from `draws_matrix` to a type that uses chain info: 

https://github.com/stan-dev/cmdstanr/issues/523

We can work around this in CmdStanR but just wanted to mention it here.

**Comment by paul-buerkner** [comment]
I see. This is a reasonable application. But the question remains how the chains would be represented inside the `draws_matrix` format?

**Comment by mjskay** [comment]
We could add number of chains as an attribute. It would be probably a annoying but not as annoying as it was for rvars :). I think one issue would be figuring out when to drop chain information depending on what kind of slicing people are doing with the matrix.

**Comment by jgabry** [comment]
> We could add number of chains as an attribute

**Comment by jgabry** [comment]
> We could add number of chains as an attribute.

I can't think of a better solution than this. 

> I think one issue would be figuring out when to drop chain information depending on what kind of slicing people are doing with the matrix.

For the CmdStanR issue I linked to above this isn't an issue because we only need to be able to convert between a full `draws_matrix` (not a sliced one) and the other formats, but yeah this seems like the major question that needs to be resolved for use cases other than the CmdStanR one. 

What if we just keep the attribute regardless of how the object is sliced?

**Comment by paul-buerkner** [comment]
I agree that we should use an attribute. 

When subsetting via `subset_draws`, it is easy to update the chain info. 

When using `[]`, it is indeed more complicated. @mjskay can we copy the logic used for `rvars` for this purpose? I don't remember right now how you did it exactly?

**Comment by mjskay** [comment]
With rvars you can't slice the draws dimension directly so this wasn't really a problem (of course someone could modify the underlying draws array directly and slice however they want but I assume if you're doing that you should know what you're doing). 

If two rvars with different numbers of chains are combined a warning is thrown and chains are collapsed.

I think the slicing logic with draws_matrix could be something like: if someone is not slicing the draws dimension then keep chain info, and if they are then drop chain info. You can't keep it when they slice the draws dimension because they might slice that dimension in a way that chains are no longer laid out in order in memory (effectively changing what chain some iterations are associated with).

**Comment by jgabry** [comment]
> I think the slicing logic with draws_matrix could be something like: if someone is not slicing the draws dimension then keep chain info, and if they are then drop chain info

That seems reasonable 



> You can't keep it when they slice the draws dimension because they might slice that dimension in a way that chains are no longer laid out in order in memory (effectively changing what chain some iterations are associated with).

Good point!

**Comment by paul-buerkner** [comment]
I have assigned this to myself and will make a PR for it.

**Comment by paul-buerkner** [comment]
Implemented via #170
